ord_schema/dataset_templating.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2020 Open Reaction Database Project Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script and functions for enumerating a dataset using a template reaction and
+a spreadsheet file.
+
+The templating code has specific expectations for how the reaction pbtxt and
+spreadsheet are defined, namely that placeholder values in the pbtxt begin and
+end with a "$" (dollar sign) and that these match a unique column header in the
+spreadsheet file.
+
+Example usage:
+
+* For normal dataset generation from a template reaction:
+  $ python dataset_templating.py --input_template=my_reaction.pbtxt
+      --input_spreadsheet=my_experiments.csv --output=my_dataset.pbtxt
+"""
+
+import os
+import re
+import pandas as pd
+
+from absl import app
+from absl import flags
+from absl import logging
+
+from google.protobuf import text_format
+
+from ord_schema import validations
+from ord_schema import message_helpers
+from ord_schema.proto import dataset_pb2
+from ord_schema.proto import reaction_pb2
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('input_template', None,
+                    'Path to a Reaction pbtxt file defining a template.')
+flags.DEFINE_string(
+    'input_spreadsheet', None,
+    'Path to a spreadsheet file (with a header row) defining '
+    'values to replace placeholders in the input_template.')
+flags.DEFINE_string('output', None, 'Filename for output Dataset.')
+flags.DEFINE_boolean('validate', True, 'If True, validate Reaction protos.')
+
+
+def read_spreadsheet(file_name):
+    """Reads a spreadsheet file."""
+    _, suffix = os.path.splitext(file_name)
+    if suffix in ['xls', 'xlsx']:
+        return pd.read_excel(file_name, dtype=str)
+    return pd.read_csv(file_name, dtype=str)
+
+
+def _replace(string, substitutions):
+    pattern = re.compile('|'.join(map(re.escape, substitutions.keys())))
+    return pattern.sub(lambda match: substitutions[match.group(0)], string)
+
+
+def generate_dataset(template_string, df, validate=True):
+    """Generates a Dataset from a template reaction formatted as text according
+    to entries in a pd.Dataframe.
+
+    Inputs:
+        template_string: The contents of a Reaction pbtxt where placeholder
+            values to be replaced are defined between dollar sign. For example,
+            a SMILES identifier value could be "$product_smiles$". Spaces
+            are not allowed.
```

Comment:
Would it be more precise to say, "placeholders may use only letters, numbers, and underscores"?